File: src/interpreter/tokenizer.rs
enum State {
    Start,
    String,
    Symbol,
}

#[derive(Clone, Copy, PartialEq, Debug)]
pub struct Location {
    pub col: usize,
    pub row: usize,
}

#[derive(Clone, PartialEq, Debug)]
pub enum Token {
    Text(String),
    Symbol(String),
    Key(String),
    NewLine,
    Equals,
    Comma,
    ObjectStart,
    ObjectEnd,
}

#[derive(Debug)]
pub struct TokenContainer {
    pub location: Location,
    pub token: Token,
}
impl TokenContainer {
    fn new(location: Location, token: Token) -> Self {
        TokenContainer { location, token }
    }
}

pub fn generate_tokens(text: &str) -> Vec<TokenContainer> {
    let mut tokens: Vec<TokenContainer> = vec![];
    let mut parser_state = State::Start;
    let mut current_stack = String::new();

    let mut location = Location { col: 1, row: 1 };

    for ch in text.chars() {
        if ch == '\n' {
            location.row += 1;
            location.col = 1;
        } else {
            location.col += 1;
        }

        parser_state = match parser_state {
            State::Start => start_state(ch, &mut location, &mut current_stack, &mut tokens),
            State::String => string_state(ch, &mut location, &mut current_stack, &mut tokens),
            State::Symbol => symbol_state(ch, &mut location, &mut current_stack, &mut tokens),
        }
    }

    if let Some(TokenContainer { token, .. }) = tokens.last() {
        if let Token::NewLine = token {
        } else {
            tokens.push(TokenContainer::new(location, Token::NewLine));
        }
    }

    tokens
}

fn start_state(
    ch: char,
    location: &mut Location,
    current_stack: &mut String,
    tokens: &mut Vec<TokenContainer>,
) -> State {
    match ch {
        '"' => {
            return State::String;
        }
        '\n' => {
            tokens.push(TokenContainer::new(*location, Token::NewLine));
            return State::Start;
        }
        '=' => {
            tokens.push(TokenContainer::new(*location, Token::Equals));
            return State::Start;
        }
        '{' => {
            tokens.push(TokenContainer::new(*location, Token::ObjectStart));
            return State::Start;
        }
        '}' => {
            tokens.push(TokenContainer::new(*location, Token::ObjectEnd));
            return State::Start;
        }
        ',' => {
            tokens.push(TokenContainer::new(*location, Token::Comma));
            return State::Start;
        }
        _ => {
            if ch.is_whitespace() {
                return State::Start;
            } else if ch.is_alphabetic() {
                current_stack.push(ch);
                return State::Symbol;
            } else {
                panic!(
                    "Tokenizer error at (row: {}, col: {})",
                    location.row, location.col
                );
            }
        }
    }
}

fn string_state(
    ch: char,
    location: &mut Location,
    current_stack: &mut String,
    tokens: &mut Vec<TokenContainer>,
) -> State {
    match ch {
        '"' => {
            tokens.push(TokenContainer::new(
                *location,
                Token::Text(current_stack.clone()),
            ));
            current_stack.clear();
            return State::Start;
        }
        _ => {
            current_stack.push(ch);
            return State::String;
        }
    }
}

fn symbol_state(
    ch: char,
    location: &mut Location,
    current_stack: &mut String,
    tokens: &mut Vec<TokenContainer>,
) -> State {
    if ch.is_alphanumeric() || ch == '_' || ch == '.' {
        current_stack.push(ch);
        return State::Symbol;
    } else if ch == '\n' {
        tokens.push(TokenContainer::new(
            *location,
            Token::Symbol(current_stack.clone()),
        ));
        tokens.push(TokenContainer::new(*location, Token::NewLine));
        current_stack.clear();
        return State::Start;
    } else if ch.is_whitespace() {
        tokens.push(TokenContainer::new(
            *location,
            Token::Symbol(current_stack.clone()),
        ));
        current_stack.clear();
        return State::Start;
    } else if ch == ':' {
        tokens.push(TokenContainer::new(
            *location,
            Token::Key(current_stack.clone()),
        ));
        current_stack.clear();
        return State::Start;
    } else if ch == ',' {
        tokens.push(TokenContainer::new(
            *location,
            Token::Symbol(current_stack.clone()),
        ));
        tokens.push(TokenContainer::new(*location, Token::Comma));
        current_stack.clear();
        return State::Start;
    } else {
        panic!(
            "Tokenizer error at location: (row: {}, col: {})",
            location.row, location.col
        );
    }
}

File: src/interpreter/tokenizer.ts
enum State {
  Start,
  String,
  Symbol,
}

interface NewLine {
  kind: "NewLine";
}
interface Text {
  kind: "Text";
  text: string;
}
interface Symbol {
  kind: "Symbol";
  text: string;
}
export type Token = NewLine | Text | Symbol;

const NewLine = (): Token => ({ kind: "NewLine" });
const Text = (text: string): Token => ({ kind: "Text", text });
const Symbol = (text: string): Token => ({ kind: "Symbol", text });

type ParserState = { state: State; stack: string; tokens: Token[] };

const startState = (parserState: ParserState, ch: string): ParserState => {
  if (ch === '"') {
    return { ...parserState, state: State.String };
  } else if (ch.match(/[\n]/)) {
    return { ...parserState, tokens: [...parserState.tokens, NewLine()] };
  } else if (ch.match(/\s/)) {
    return parserState;
  } else if (ch.match(/[a-zA-Z_]/)) {
    return {
      ...parserState,
      stack: parserState.stack + ch,
      state: State.Symbol,
    };
  } else {
    throw "Tokenizer error";
  }
};

const stringState = (parserState: ParserState, ch: string): ParserState => {
  if (ch == '"') {
    return {
      tokens: [...parserState.tokens, Text(parserState.stack)],
      stack: "",
      state: State.Start,
    };
  } else {
    return { ...parserState, stack: parserState.stack + ch };
  }
};

const symbolState = (parserState: ParserState, ch: string): ParserState => {
  if (ch.match(/[\w.]/)) {
    return { ...parserState, stack: parserState.stack + ch };
  } else if (ch.match(/[\n]/)) {
    return {
      tokens: [...parserState.tokens, Symbol(parserState.stack), NewLine()],
      stack: "",
      state: State.Start,
    };
  } else if (ch.match(/\s/)) {
    return {
      tokens: [...parserState.tokens, Symbol(parserState.stack)],
      stack: "",
      state: State.Start,
    };
  } else {
    throw "Tokenizer error";
  }
};

export const generateTokens = (text: string): Token[] => {
  let parserState: ParserState = {
    state: State.Start,
    stack: "",
    tokens: [],
  };

  for (const ch of text + "\n") {
    switch (parserState.state) {
      case State.Start:
        parserState = startState(parserState, ch);
        break;
      case State.String:
        parserState = stringState(parserState, ch);
        break;
      case State.Symbol:
        parserState = symbolState(parserState, ch);
        break;
    }
  }

  return parserState.tokens;
};
